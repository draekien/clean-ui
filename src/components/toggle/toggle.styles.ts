import { ThemeUIStyleObject } from '@theme-ui/css';

export interface ToggleCssProps {
  checked?: boolean;
  disabled?: boolean;
}

export const toggleCss = ({ checked, disabled }: ToggleCssProps): ThemeUIStyleObject => ({
  height: '2em',
  width: '4em',
  marginX: 'xs',
  borderRadius: 'sm',
  border: '0.25em solid',
  borderColor: 'border',
  position: 'relative',
  backgroundColor: disabled ? 'b-300' : 'border',
  boxShadow: 'inset 0 0 1em rgba(61, 61, 61, 0.25)',
  cursor: disabled ? 'not-allowed' : 'pointer',
  ':hover, :focus': {
    ':before': {
      backgroundColor: disabled ? 'b-200' : 'b-000',
    },
  },
  ':after': {
    content: '""',
    position: 'absolute',
    top: 0,
    left: 0,
    height: '100%',
    width: checked ? '100%' : '0%',
    transition: 'all 100ms ease-in-out',
    backgroundColor: disabled ? 'success-100' : 'success-300',
    boxShadow: 'inset 0 0 1em rgba(61, 61, 61, 0.25)',
    zIndex: 'low',
  },
  ':before': {
    content: '""',
    position: 'absolute',
    boxSizing: 'content-box',
    top: 0,
    left: checked ? '55%' : '0%',
    height: '100%',
    width: '45%',
    borderRadius: 'xs',
    backgroundColor: disabled ? 'b-200' : 'b-100',
    transition: 'all 100ms ease-in-out',
    zIndex: 'middle',
  },
});

export const toggleContainerCss: ThemeUIStyleObject = {
  display: 'flex',
  flexFlow: 'row nowrap',
  alignItems: 'center',
};
