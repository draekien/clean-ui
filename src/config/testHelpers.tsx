import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: ReactNode) =>
  render(<ThemeProvider>{element}</ThemeProvider>);
