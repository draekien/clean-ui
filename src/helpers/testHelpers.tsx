import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../components/theme/theme.provider';

export const themedRender = (element: React.ReactNode) => {
  return render(<ThemeProvider>{element}</ThemeProvider>);
};

export const hexToRgb = (hex: string) => {
  hex = hex.replace(/[^0-9A-F]/gi, '');
  const bigint = parseInt(hex, 16);
  const r = (bigint >> 16) & 255;
  const g = (bigint >> 8) & 255;
  const b = bigint & 255;

  return `rgb(${r}, ${g}, ${b})`;
};
