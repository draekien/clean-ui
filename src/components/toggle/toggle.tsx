/** @jsxImportSource theme-ui */
import React, { forwardRef, useState, useEffect } from 'react';
import { Label } from '../label/label';
import { toggleContainerCss, toggleCss } from './toggle.styles';

export interface ToggleProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'onClick'> {
  inputId: string;
  on?: React.ReactNode;
  off?: React.ReactNode;
  onClick?: (checked: boolean) => void;
}

export const Toggle: React.ForwardRefExoticComponent<
  ToggleProps & React.RefAttributes<HTMLInputElement>
> = forwardRef(({ ref, on, off, inputId, disabled, onClick, ...rest }) => {
  const [checked, setChecked] = useState(rest.checked);

  useEffect(() => {
    setChecked(rest.checked);
  }, [rest.checked]);

  const handleClick = () => {
    onClick && onClick(!checked);
    setChecked(!checked);
  };

  return (
    <div>
      <input
        {...rest}
        ref={ref}
        id={inputId}
        name={inputId}
        type="checkbox"
        checked={checked}
        disabled={disabled}
        hidden
      />
      <Label htmlFor={inputId}>
        <div sx={toggleContainerCss}>
          <span>{off}</span>
          <button
            sx={toggleCss({ checked, disabled })}
            onClick={handleClick}
            disabled={disabled}
          />
          <span>{on}</span>
        </div>
      </Label>
    </div>
  );
});
