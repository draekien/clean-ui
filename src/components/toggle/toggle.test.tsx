import React from 'react';
import { Toggle } from './toggle';
import { themedRender } from '../../helpers/testHelpers';

describe('Toggle component', () => {
  it('should match snapshot', () => {
    const { container } = themedRender(<Toggle inputId="toggle" on="on" off="off" />);

    expect(container).toMatchSnapshot();
  });
});
