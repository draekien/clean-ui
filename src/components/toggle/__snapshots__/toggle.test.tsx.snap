// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle component should match snapshot 1`] = `
<div>
  <div>
    <input
      hidden=""
      id="toggle"
      name="toggle"
      type="checkbox"
    />
    <label
      class="css-1pst1pz-Label"
      for="toggle"
    >
      <div
        class="css-19odyxe"
      >
        <span>
          off
        </span>
        <button
          class="css-faixig"
        />
        <span>
          on
        </span>
      </div>
    </label>
  </div>
</div>
`;
